pySDC/implementations/controller_classes/controller_nonMPI.py

```diff
@@ -712,3 +746,61 @@ def default(self, local_MS_running):
             local_MS_running (list): list of currently running steps
         """
         raise ControllerError('Unknown stage, got %s' % local_MS_running[0].status.stage)  # TODO
+
+    def resilence(self, local_MS_running):
+        """
+        Call various functions that are supposed to provide some sort of resilience from here
+        """
+
+        if self.params.use_HotRod:
+            self.hotrod(local_MS_running)
+
+    def hotrod(self, local_MS_running):
+        """
+        See for the reference:
+        Lightweight and Accurate Silent Data Corruption Detection in Ordinary Differential Equation Solvers,
+        Guhur et al. 2016, Springer. DOI: 10.1007/978-3-319-43659-3_47
+        """
+        for i in range(len(local_MS_running)):
+            S = local_MS_running[i]
+            if S.status.iter == S.params.maxiter:
+                for l in S.levels:
+                    # throw away the final sweep to match the error estimates
+                    l.u[:] = l.uold[:]
+
+                    # check if a fault is detected
+                    if None not in [l.status.e_extrapolated, l.status.e_embedded]:
+                        diff = l.status.e_extrapolated - l.status.e_embedded
+                        if diff > self.params.HotRod_tol:
+                            self.restart[i] = True
+
+    def adaptivity(self, MS):
+        """
+        Method to compute time step size adaptively based on embedded error estimate
+        """
+
+        # loop through steps and compute local error and optimal step size from there
+        for i in range(len(MS)):
+            S = MS[i]
+
+            # check if we performed the desired amount of sweeps
+            if S.status.iter < S.params.maxiter:
+                continue
+
+            L = S.levels[0]
+
+            # compute next step size
+            order = S.status.iter  # embedded error estimate is same order as time marching
+            h_opt = L.params.dt * 0.9 * (L.params.e_tol / L.status.e_embedded)**(1. / order)
```

Comment:
Same question here as in the hotrod

Reply:
I added an assertion in case the error was not estimated before calling adaptivity, which might happen when we start doing PinT adaptivity, but that is really something to be changed in the error_estimation module, which I don't think should be changed by the user later on. So it's really just for me. I actually raise a parameter error when e_tol is not specified since that's equivalent to no supplying a step size in a fixed step scheme and I don't think there is a universally useful choice. I can ofc just put in like 1e-5 but I think forcing the user to make a choice themselves is better.